collections/accounts.go

```diff
@@ -0,0 +1,526 @@
+	keys := make([][]byte, len(members))
+	var err error
+	for i, m := range members {
+		keys[i], err = crypto.MarshalPublicKey(m.Key)
+		if err != nil {
+			return nil, err
+		}
+	}
+	cursor, err := a.col.Find(ctx, bson.M{"_id": bson.M{"$in": keys}})
+	if err != nil {
+		return nil, err
+	}
+	var docs []Account
+	for cursor.Next(ctx) {
+		var raw bson.M
+		if err := cursor.Decode(&raw); err != nil {
+			return nil, err
+		}
+		doc, err := decodeAccount(raw)
+		if err != nil {
+			return nil, err
+		}
+		docs = append(docs, *doc)
+	}
+	if err := cursor.Err(); err != nil {
+		return nil, err
+	}
+	return docs, nil
+}
+
+func (a *Accounts) IsOwner(ctx context.Context, username string, member crypto.PubKey) (bool, error) {
+	mid, err := crypto.MarshalPublicKey(member)
+	if err != nil {
+		return false, err
+	}
+	filter := bson.M{"username": username, "members": bson.M{"$elemMatch": bson.M{"_id": mid, "role": OrgOwner}}}
+	res := a.col.FindOne(ctx, filter)
+	if res.Err() != nil {
+		if errors.Is(res.Err(), mongo.ErrNoDocuments) {
+			return false, nil
+		} else {
+			return false, res.Err()
+		}
+	}
+	return true, nil
+}
+
+func (a *Accounts) IsMember(ctx context.Context, username string, member crypto.PubKey) (bool, error) {
+	mid, err := crypto.MarshalPublicKey(member)
+	if err != nil {
+		return false, err
+	}
+	filter := bson.M{"username": username, "members": bson.M{"$elemMatch": bson.M{"_id": mid}}}
+	res := a.col.FindOne(ctx, filter)
+	if res.Err() != nil {
+		if errors.Is(res.Err(), mongo.ErrNoDocuments) {
+			return false, nil
+		} else {
+			return false, res.Err()
+		}
+	}
+	return true, nil
+}
+
+func (a *Accounts) AddMember(ctx context.Context, username string, member Member) error {
+	mk, err := crypto.MarshalPublicKey(member.Key)
+	if err != nil {
+		return err
+	}
+	raw := bson.M{
+		"_id":      mk,
+		"username": member.Username,
+		"role":     int(member.Role),
+	}
+	_, err = a.col.UpdateOne(ctx, bson.M{"username": username, "members._id": bson.M{"$ne": mk}}, bson.M{"$push": bson.M{"members": raw}})
+	return err
+}
+
+func (a *Accounts) RemoveMember(ctx context.Context, username string, member crypto.PubKey) error {
+	isOwner, err := a.IsOwner(ctx, username, member)
+	if err != nil {
+		return err
+	}
+	if isOwner { // Ensure there will still be at least one owner left
+		cursor, err := a.col.Aggregate(ctx, mongo.Pipeline{
+			bson.D{{"$match", bson.M{"username": username}}},
+			bson.D{{"$project", bson.M{
+				"members": bson.M{
+					"$filter": bson.M{
+						"input": "$members",
+						"as":    "member",
+						"cond":  bson.M{"$eq": bson.A{"$$member.role", 0}},
+					},
+				},
+			}}},
+			bson.D{{"$count", "members"}},
+		}, options.Aggregate().SetHint(bson.D{{"username", 1}}))
+		if err != nil {
+			return err
+		}
+		type res struct {
+			Owners int `bson:"members"`
+		}
+		var r res
+		for cursor.Next(ctx) {
+			if err := cursor.Decode(&r); err != nil {
+				return err
+			}
+			break
+		}
```

Comment:
Maybe the for look can be avoided since it always terminates after first Next?

Reply:
Good call... done.